Codeql 2.16.3 changelogs (#15750)
github/codeql: docs/codeql/codeql-overview/codeql-changelog/codeql-cli-2.16.3.rst

.. contents:: Contents
   :depth: 2
   :local:
   :backlinks: none

This is an overview of changes in the CodeQL CLI and relevant CodeQL query and library packs. For additional updates on changes to the CodeQL code scanning experience, check out the `code scanning section on the GitHub blog <https://github.blog/tag/code-scanning/>`__, `relevant GitHub Changelog updates <https://github.blog/changelog/label/code-scanning/>`__, `changes in the CodeQL extension for Visual Studio Code <https://marketplace.visualstudio.com/items/GitHub.vscode-codeql/changelog>`__, and the `CodeQL Action changelog <https://github.com/github/codeql-action/blob/main/CHANGELOG.md>`__.

Security Coverage
-----------------

CodeQL 2.16.3 runs a total of 408 security queries when configured with the Default suite (covering 160 CWE). The Extended suite enables an additional 131 queries (covering 34 more CWE). 2 security queries have been added with this release.

CodeQL CLI
----------

Bug Fixes
~~~~~~~~~

*   Fixed a bug where CodeQL may produce an invalid database when it exhausts all available ID numbers. Now it detects the condition and reports an error instead.

New Features
~~~~~~~~~~~~

> *   A new extractor option has been added to the Python extractor.
>     Pass one of :code:`--extractor-option python_executable_name=py` or :code:`--extractor-option python_executable_name=python` or :code:`--extractor-option python_executable_name=python3` to :code:`codeql database create` (or :code:`codeql database trace-command` or,
>     for indirect tracing, :code:`codeql database init`) to override the default Python executable search and selection behavior of the Python extractor. For example, on Windows machines, the Python extractor will expect to find :code:`py.exe` on the system :code:`PATH` by default.
>     Setting this extractor option or environment variable allows overriding this behavior to look for a different name.

Reviewer: It looks as if `--extractor-option` already exists and that the new option is actually `python_executable_name`. This doesn't stand out with the current description.

I'd suggest something more like:

```suggestion
*   A new extractor option has been added to the Python extractor: :code:`python_executable_name`.
    You can use this option to override the default process the extractor uses to find and select a Python executable.
    Pass one of :code:`--extractor-option python_executable_name=py` or :code:`--extractor-option python_executable_name=python` or :code:`--extractor-option python_executable_name=python3` to commands that run the extractor, for example: :code:`codeql database create`. 
    
    On Windows machines, the Python extractor will expect to find :code:`py.exe` on the system :code:`PATH` by default.
    If the Python extractor has a different name, you can set the new extractor option, or environment variable, to override this value and look for :code:`python.exe` or :code:`python3.exe`.
```

Reviewer 2: > If the Python extractor has a different name

s/extractor/executable, and I would drop mention of the environment variable since we're not mentioning it above

Otherwise agree!

Reviewer: Thank you for saving my blushes 🙇🏻‍♀️ 

Suggestion updated accordingly.

>     More detail can be found in `the extractor option documentation <https://docs.github.com/en/code-security/codeql-cli/using-the-advanced-functionality-of-the-codeql-cli/extractor-options>`__.

Reviewer: I was expecting to find more information about the new Python option when I clicked this link. I suggest that we're more explicit about what people will find. Maybe:

```suggestion
    For more information about using the extractor option with the CodeQL CLI, see `Extractor options <https://docs.github.com/en/code-security/codeql-cli/using-the-advanced-functionality-of-the-codeql-cli/extractor-options>`__.
```

Java
""""

*   The sinks of the queries :code:`java/path-injection` and :code:`java/path-injection-local` have been reworked. Path creation sinks have been converted to summaries instead, while sinks now are actual file read/write operations only. This has reduced the false positive ratio of both queries.

Minor Analysis Improvements
~~~~~~~~~~~~~~~~~~~~~~~~~~~

C/C++
"""""

*   The "non-constant format string" query (:code:`cpp/non-constant-format`) has been updated to produce fewer false positives.
*   Added dataflow models for the :code:`gettext` function variants.

C#
""

*   Added sanitizers for relative URLs, :code:`List.Contains()`, and checking the :code:`.Host` property on an URI to the :code:`cs/web/unvalidated-url-redirection` query.

Java
""""

*   The sanitizer for the path injection queries has been improved to handle more cases where :code:`equals` is used to check an exact path match.

> *   The query :code:`java/unvalidated-url-redirection` now sanitizes results following the same logic as the query :code:`java/ssrf`. URLs the destination of which cannot be externally controlled will not be reported anymore.

Reviewer: Suggest perhaps:

```suggestion
*   The query :code:`java/unvalidated-url-redirection` now sanitizes results following the same logic as the query :code:`java/ssrf`. URLs where the destination cannot be controlled externally are no longer reported.
```
